// vitepress-plugin-twoslash@0.10.0/dist/client.mjs
import FloatingVue, { recomputeAllPoppers } from 'floating-vue';

const isMobile = typeof navigator !== "undefined" && /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent);
const TwoslashFloatingVue = {
  install: (app, options = {}) => {
    if (typeof window !== "undefined") {
      window.addEventListener("click", (e) => {
        const path = e.composedPath();
        if (path.some((el) => el?.classList?.contains?.("vp-code-group") || el?.classList?.contains?.("tabs")))
          recomputeAllPoppers();
      }, { passive: true });
    }
    app.use(FloatingVue, {
      ...options,
      themes: {
        twoslash: {
          $extend: "dropdown",
          triggers: isMobile ? ["touch"] : ["hover", "touch"],
          popperTriggers: isMobile ? ["touch"] : ["hover", "touch"],
          overflowPadding: 10,
          delay: 0,
          handleResize: false,
          autoHide: true,
          instantMove: true,
          flip: false
        },
        ...options.theme
      }
    });
  }
};

export { TwoslashFloatingVue as default };
